pkg/grpc/labelstore/client/client.go

```diff
@@ -51,15 +59,31 @@ func (c Client) WatchMatches(selector klabels.Selector, labelType labels.Type, q
 		for {
 			labeled, err := watchClient.Recv()
 			if grpc.Code(err) == codes.Canceled {
+				c.logger.Infoln("label store client: terminating WatchMatches()")
 				// This just means quitCh fired and the RPC was canceled as expected
 				return
 			}
 
 			if err != nil {
-				// doesn't hurt to call cancelFunc() more than once potentially
-				cancelFunc()
-				c.logger.WithError(err).Errorln("Unexpected error reading from WatchMatches stream")
-				return
+				innerCancel()
+				c.logger.WithError(err).Errorln("unexpected error reading from WatchMatches stream, starting another RPC")
+
+				watchClient = nil
+
+				for watchClient == nil {
+					innerCtx, innerCancel = context.WithCancel(outerCtx)
+
+					time.Sleep(2 * time.Second)
```

Comment:
I share your aversion, although i'm not sure what's worse. Options are:

1) make this retry forever
2) add a case where the output channel might be closed before the caller closes `quitCh`. All callers across implementations (e.g. direct consul access) would have to be modified to detect this and start another `WatchMatches`